File: snake_game/server-skynet/lualib/snake/protocol.lua
-- Package and Message protocol implementation
-- Compatible with C# client

local M = {}

-- Package types
M.PACKAGE_TYPE = {
    HANDSHAKE = 1,
    HANDSHAKE_ACK = 2,
    HEARTBEAT = 3,
    DATA = 4,
    KICK = 5,
}

-- Message types
M.MESSAGE_TYPE = {
    REQUEST = 0,
    NOTIFY = 1,
    RESPONSE = 2,
    PUSH = 3,
}

-- Helper functions for binary encoding
local function write_byte(value)
    return string.char(value % 256)
end

local function write_int24(value)
    return string.char(
        math.floor(value / 65536) % 256,
        math.floor(value / 256) % 256,
        value % 256
    )
end

local function read_byte(buffer, offset)
    return string.byte(buffer, offset)
end

local function read_int24(buffer, offset)
    local b1 = string.byte(buffer, offset)
    local b2 = string.byte(buffer, offset + 1)
    local b3 = string.byte(buffer, offset + 2)
    return b1 * 65536 + b2 * 256 + b3
end

-- Encode package: [type(1 byte)][length(3 bytes)][body]
function M.encode_package(pkg_type, body)
    local body_len = body and #body or 0
    local data = write_byte(pkg_type) .. write_int24(body_len)
    if body_len > 0 then
        data = data .. body
    end
    return data
end

-- Decode package from buffer
function M.decode_package(buffer, offset)
    offset = offset or 1
    if #buffer < offset + 3 then
        return nil, offset
    end
    
    local pkg_type = read_byte(buffer, offset)
    local body_len = read_int24(buffer, offset + 1)
    
    if #buffer < offset + 3 + body_len then
        return nil, offset
    end
    
    local body = ""
    if body_len > 0 then
        body = string.sub(buffer, offset + 4, offset + 3 + body_len)
    end
    
    return {
        type = pkg_type,
        length = body_len,
        body = body,
    }, offset + 4 + body_len
end

-- Encode message: [flag(1 byte)][id(base128)][route][body]
function M.encode_message(id, msg_type, compress_route, route, body)
    local result = {}
    
    -- Encode flag: type(3 bits) << 1 | compressRoute(1 bit)
    local flag = (msg_type * 2) % 256
    if compress_route then
        flag = flag + 1
    end
    table.insert(result, string.char(flag))
    
    -- Encode id (base128, only for REQUEST/RESPONSE)
    if msg_type == M.MESSAGE_TYPE.REQUEST or msg_type == M.MESSAGE_TYPE.RESPONSE then
        local id_val = id
        repeat
            local tmp = id_val % 128
            local next = math.floor(id_val / 128)
            if next ~= 0 then
                tmp = tmp + 128
            end
            table.insert(result, string.char(tmp))
            id_val = next
        until id_val == 0
    end
    
    -- Encode route (only for REQUEST/NOTIFY/PUSH)
    if msg_type == M.MESSAGE_TYPE.REQUEST or msg_type == M.MESSAGE_TYPE.NOTIFY or msg_type == M.MESSAGE_TYPE.PUSH then
        if compress_route then
            -- Compressed route: 2 bytes (big-endian)
            table.insert(result, string.char(0, 0))
        else
            -- Full route string: 1 byte length + route string
            local route_bytes = route or ""
            table.insert(result, string.char(#route_bytes))
            if #route_bytes > 0 then
                table.insert(result, route_bytes)
            end
        end
    end
    
    -- Encode body
    if body and #body > 0 then
        table.insert(result, body)
    end
    
    return table.concat(result)
end

-- Decode message
function M.decode_message(data)
    if #data < 1 then
        return nil
    end
    
    local offset = 1
    
    -- Parse flag (1 byte)
    local flag = string.byte(data, offset)
    offset = offset + 1
    
    local compress_route = (flag % 2) == 1
    local msg_type = math.floor(flag / 2) % 8
    local compress_gzip = (math.floor(flag / 16) % 2) == 1
    
    -- Parse id (base128 encoded, only for REQUEST/RESPONSE)
    local id = 0
    if msg_type == M.MESSAGE_TYPE.REQUEST or msg_type == M.MESSAGE_TYPE.RESPONSE then
        local i = 0
        while true do
            if offset > #data then
                return nil
            end
            local m = string.byte(data, offset)
            id = id + ((m % 128) * (2 ^ (7 * i)))
            offset = offset + 1
            i = i + 1
            if m < 128 then
                break
            end
        end
    end
    
    -- Parse route (only for REQUEST/NOTIFY/PUSH)
    local route = ""
    if msg_type == M.MESSAGE_TYPE.REQUEST or msg_type == M.MESSAGE_TYPE.NOTIFY or msg_type == M.MESSAGE_TYPE.PUSH then
        if compress_route then
            -- Compressed route: 2 bytes (big-endian)
            if offset + 2 > #data then
                return nil
            end
            offset = offset + 2
        else
            -- Full route string: 1 byte length + route string
            if offset > #data then
                return nil
            end
            local route_len = string.byte(data, offset)
            offset = offset + 1
            if route_len > 0 then
                if offset + route_len > #data then
                    return nil
                end
                route = string.sub(data, offset, offset + route_len - 1)
                offset = offset + route_len
            end
        end
    end
    
    -- Read body (remaining bytes)
    local body = ""
    if offset <= #data then
        body = string.sub(data, offset)
    end
    
    return {
        id = id,
        type = msg_type,
        compress_route = compress_route,
        route = route,
        body = body,
        compress_gzip = compress_gzip,
    }
end

return M

File: echo/server-cpp/protocol.cpp
#include "protocol.hpp"

namespace protocol {

std::vector<uint8_t> Package::encode(PackageType type, const std::vector<uint8_t>& body) {
    size_t body_len = body.size();
    std::vector<uint8_t> result(4 + body_len);

    result[0] = static_cast<uint8_t>(type);
    result[1] = static_cast<uint8_t>((body_len >> 16) & 0xFF);
    result[2] = static_cast<uint8_t>((body_len >> 8) & 0xFF);
    result[3] = static_cast<uint8_t>(body_len & 0xFF);

    if (body_len > 0) {
        std::copy(body.begin(), body.end(), result.begin() + 4);
    }

    return result;
}

std::optional<Package> Package::decode(const std::vector<uint8_t>& data) {
    if (data.size() < 4) return std::nullopt;

    auto pkg_type = static_cast<PackageType>(data[0]);
    int length = (data[1] << 16) | (data[2] << 8) | data[3];

    if (static_cast<int>(data.size()) < 4 + length) return std::nullopt;

    Package pkg;
    pkg.type = pkg_type;
    pkg.body.assign(data.begin() + 4, data.begin() + 4 + length);

    return pkg;
}

std::vector<uint8_t> Message::encode(int id, MessageType type, bool compress_route,
                                     const std::string& route, const std::vector<uint8_t>& body) {
    std::vector<uint8_t> result;

    // Encode flag
    uint8_t flag = static_cast<uint8_t>(static_cast<int>(type) << 1);
    if (compress_route) flag |= 1;
    result.push_back(flag);

    // Encode id (base128, only for REQUEST/RESPONSE)
    if (type == MessageType::Request || type == MessageType::Response) {
        int id_val = id;
        do {
            int tmp = id_val % 128;
            int next = id_val / 128;
            if (next != 0) tmp += 128;
            result.push_back(static_cast<uint8_t>(tmp));
            id_val = next;
        } while (id_val != 0);
    }

    // Encode route (only for REQUEST/NOTIFY/PUSH)
    if (type == MessageType::Request || type == MessageType::Notify || type == MessageType::Push) {
        if (compress_route) {
            result.push_back(0);
            result.push_back(0);
        } else {
            result.push_back(static_cast<uint8_t>(route.size()));
            result.insert(result.end(), route.begin(), route.end());
        }
    }

    // Encode body
    result.insert(result.end(), body.begin(), body.end());

    return result;
}

std::optional<Message> Message::decode(const std::vector<uint8_t>& data) {
    if (data.empty()) return std::nullopt;

    size_t offset = 0;

    // Parse flag
    uint8_t flag = data[offset++];
    bool compress_route = (flag & 0x1) == 1;
    auto msg_type = static_cast<MessageType>((flag >> 1) & 0x7);
    bool compress_gzip = ((flag >> 4) & 0x1) == 1;

    // Parse id (base128, only for REQUEST/RESPONSE)
    int id = 0;
    if (msg_type == MessageType::Request || msg_type == MessageType::Response) {
        int i = 0;
        while (true) {
            if (offset >= data.size()) return std::nullopt;
            uint8_t m = data[offset++];
            id += (m & 0x7F) << (7 * i);
            i++;
            if (m < 128) break;
        }
    }

    // Parse route
    std::string route;
    if (msg_type == MessageType::Request || msg_type == MessageType::Notify || msg_type == MessageType::Push) {
        if (compress_route) {
            if (offset + 2 > data.size()) return std::nullopt;
            offset += 2;
        } else {
            if (offset >= data.size()) return std::nullopt;
            int route_len = data[offset++];
            if (route_len > 0) {
                if (offset + route_len > data.size()) return std::nullopt;
                route.assign(data.begin() + offset, data.begin() + offset + route_len);
                offset += route_len;
            }
        }
    }

    // Read body
    std::vector<uint8_t> body;
    if (offset < data.size()) {
        body.assign(data.begin() + offset, data.end());
    }

    Message msg;
    msg.id = id;
    msg.type = msg_type;
    msg.compress_route = compress_route;
    msg.route = route;
    msg.body = body;
    msg.compress_gzip = compress_gzip;

    return msg;
}

} // namespace protocol

